cassiopeia/data/CassiopeiaTree.py

```diff
@@ -0,0 +1,289 @@
+queried like the average tree depth or agreement between character states and
+phylogeny.
+
+This object can be passed to any CassiopeiaSolver subclass as well as any
+analysis module, like a branch length estimator or rate matrix estimator
+"""
+import ete3
+import networkx as nx
+import pandas as pd
+from typing import Dict, List, Optional, Union
+
+from cassiopeia.data import utilities
+
+
+class CassiopeiaTree:
+    """Basic tree object for Cassiopeia.
+
+    This object stores the key attributes and functionalities a user might want
+    for working with lineage tracing experiments. At its core, it stores
+    three main items - a tree, a character matrix, and meta data associated
+    with the data.
+
+    The tree can be fed into the object via Ete3, Networkx, or can be inferred
+    using one of the CassiopeiaSolver algorithms in the `solver` module.
+
+    A character matrix can be stored in the object, containing the states
+    observed for each cell. In typical lineage tracing experiments, these are
+    integer representations of the indels observed at each unique cut site.
+
+    Meta data for cells or characters can also be stored in this object. These
+    items can be categorical or numerical in nature. Common examples of cell
+    meta data are the cluster identity, tissue identity, or number of target-site
+    UMIs per cell. These items can be used in downstream analyses, for example
+    the FitchCount algorithm which infers the number of transitions between
+    categorical variables (e.g., tissues). Common examples of character meta
+    data are the proportion of missing data for each character or the entropy
+    of states. These are good statistics to have for feature selection.
+
+    TODO(mattjones315): Add experimental meta data as arguments.
+
+    Args:
+        character_matrix: The character matrix for the lineage.
+        cell_meta: Per-cell meta data
+        character_meta: Per-character meta data
+        priors: A dictionary storing the probability of a character mutating
+            to a particular state.
+        tree: A tree for the lineage. 
+    """
+
+    def __init__(
+        self,
+        character_matrix: Optional[pd.DataFrame] = None,
+        cell_meta: Optional[pd.DataFrame] = None,
+        character_meta: Optional[pd.DataFrame] = None,
+        priors: Optional[Dict[int, Dict[int, float]]] = None,
+        tree: Optional[Union[str, ete3.Tree, nx.DiGraph]] = None
+    ):
+
+        self.character_matrix = character_matrix
+        self.cell_meta = cell_meta
+        self.character_meta = character_meta
+        self.priors = priors
+        self.__network = None
+
+        if isinstance(tree, nx.DiGraph):
+            self.__network = tree
+        if isinstance(tree, str):
+            self.__network = utilities.newick_to_networkx(tree)
+        if isinstance(tree, ete3.Tree):
+            self.__network = utilities.ete3_to_networkx(tree)
+
+    @property
```

Comment:
For right now, I don't have functions for mutating the tree structure. 

I totally agree though - the caching of properties will be come quite troublesome once we have methods for pruning, collapsing, etc.

Reply:
Yeah, and this reminds me that for branch length estimation I'll have to resolve multifurcations, which begs the question where will this happen. It could be in CassiopeiaTree, but then we have to keep a bunch of stuff up-to-date... or the CassiopeiaSolver could do it by request, or a new module/function, etc. TBD.

Reply:
This is a good point, and we should be sure to flag it when developing something like the Branch Length Estimator. 

I've added it as a docstring todo.